drivers: spi: Add Intel SEDI driver

yeweizy opened the pull request:
Adds a new SPI shim driver for Intel SoCs. Builds upon the SEDI bare metal SPI driver in the hal-intel module.

Originally authored by Kong Li.

likongintel commented on dts/bindings/spi/intel,sedi-spi.yaml:
```diff
@@ -0,0 +1,18 @@
+#
+# Copyright (c) 2019 - 2021 Intel Corporation.
```
Copyright

kwd-doodling commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+	pm_device_busy_set(dev);
+
+	/* If device locked, will return error, need to release first */
+	if ((spi->is_locked) && (!spi_context_configured(&spi->ctx, config))) {
```
please revise all is_locked logics.
it seems the device locking SPI_LOCK_ON is totally handled by spi_context.h now.

yeweizy replied:
Will do.

kwd-doodling commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+	struct spi_context *ctx = &spi->ctx;
+	int error;
+
+	/* Update the semaphore of spi context */
```
?
please check all other comments, correctness or necessary. remove the ones if the code can explain itself.

yeweizy replied:
will do

kwd-doodling commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+	struct spi_sedi_data *spi = dev->data;
+	int ret;
+
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
```
to make this work with paging enabled, need to add DEVICE_MMIO_RAM and DEVICE_MMIO_ROM to spi_sedi_config and spi_sedi_data, see sedi_i2c.c.

yeweizy replied:
Will do.

kwd-doodling commented on drivers/spi/Kconfig.sedi:
```diff
@@ -0,0 +1,14 @@
+	bool "SEDI SPI controller driver"
+	default y
+	depends on DT_HAS_INTEL_SEDI_SPI_ENABLED
+	help
```
make similar as other SEDI drivers, see drivers/i2c/Kconfig.sedi and drivers/gpio/Kconfig.sedi

yeweizy replied:
will do

kwd-doodling commented on drivers/spi/Kconfig.sedi:
```diff
@@ -0,0 +1,14 @@
+#
+
+config SPI_SEDI
+	bool "SEDI SPI controller driver"
```
add Intel

yeweizy replied:
will do.

yeweizy commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+#define DT_DRV_COMPAT intel_sedi_spi
+
+#include "sedi_driver_spi.h"
+#include "sedi_driver_dma.h"
```
will fix.

yeweizy commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+
+#define DT_DRV_COMPAT intel_sedi_spi
+
+#include "sedi_driver_spi.h"
```
will do.

yeweizy commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+#include "sedi_driver_dma.h"
+#define LOG_LEVEL CONFIG_SPI_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(spi_sedi);
```
There is LOG_xxx() in "spi_context.h" which included

yeweizy commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+
+#define SPI_NOT_USE_DMA (-1)
+#define SPI_LINE_MODE_OFF (16)
+#define SPI_DMA_THRESHOLD (32)
```
will do.

yeweizy commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,435 @@
+	const struct spi_buf *buf;
+	bool is_multibufs = false;
+
+	if (config != NULL) {
```
It's referred from our internal source code.  
A valid call will definitely have this config pointer as a not NULL value, so always not necessary, will remove it.

kwd-doodling commented on dts/x86/intel/intel_ish5.dtsi:
```diff
@@ -163,5 +163,28 @@
 
 			status = "okay";
 		};
```
add one blank line.

yeweizy replied:
will do

yeweizy committed:
drivers: spi: Add Intel SEDI driver

Add a new SPI shim driver for Intel SoCs. Builds upon the SEDI bare
metal SPI driver in the hal-intel module.

Co-Authored-By: Kong Li <li.kong@intel.com>
Signed-off-by: Ye Weize <weize.ye@intel.com>

teburd commented on drivers/spi/spi_sedi.c:
```diff
@@ -0,0 +1,415 @@
+	struct spi_context *ctx = &spi->ctx;
+	int error;
+
+	if (event == SEDI_SPI_EVENT_DATA_LOST) {
```
Seems like this branching over event could easily be done using a switch, even for the error case, not really a big deal the way its written but might be a little nicer

yeweizy replied:
Yes Thanks Teburd for the comment. A switch could be a little better, but it won't make much difference here because there are only 2 different cases. I will choose not to push another patch to replace it this time.
